RuntimeException: Unable to stop activity {ui.ReadArticleActivity}: SQLiteConstraintException: UNIQUE constraint failed: ARTICLE.article_id (code 2067)

imported from gdev console:

```
java.lang.RuntimeException: Unable to stop activity {fr.gaulupeau.apps.InThePoche/fr.gaulupeau.apps.Poche.ui.ReadArticleActivity}: android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: ARTICLE.article_id (code 2067)
	at android.app.ActivityThread.performStopActivityInner(ActivityThread.java:3837)
	at android.app.ActivityThread.handleStopActivity(ActivityThread.java:3886)
	at android.app.ActivityThread.-wrap25(ActivityThread.java)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1494)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6077)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
Caused by: android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: ARTICLE.article_id (code 2067)
	at android.database.sqlite.SQLiteConnection.nativeExecute(Native Method)
	at android.database.sqlite.SQLiteConnection.execute(SQLiteConnection.java:555)
	at android.database.sqlite.SQLiteSession.execute(SQLiteSession.java:619)
	at android.database.sqlite.SQLiteStatement.execute(SQLiteStatement.java:44)
	at de.greenrobot.dao.AbstractDao.updateInsideSynchronized(AbstractDao.java:725)
	at de.greenrobot.dao.AbstractDao.update(AbstractDao.java:700)
	at fr.gaulupeau.apps.Poche.ui.ReadArticleActivity.onStop(ReadArticleActivity.java:638)
	at android.app.Instrumentation.callActivityOnStop(Instrumentation.java:1289)
	at android.app.Activity.performStop(Activity.java:6839)
	at android.app.ActivityThread.performStopActivityInner(ActivityThread.java:3834)
	... 9 more
```

Android 7.0 with Nexus 5X (bullhead) app version 29
Android 5.1 with Nexus 4 (mako) on a. v. 28
Android 4.4 with wiko on a.v. 27

reported 5 times in total

--- a/app/src/main/java/fr/gaulupeau/apps/Poche/ui/ReadArticleActivity.java
+++ b/app/src/main/java/fr/gaulupeau/apps/Poche/ui/ReadArticleActivity.java
@@ -74,6 +74,7 @@ public class ReadArticleActivity extends BaseActionBarActivity {
     private View hrBar;
     private TtsFragment ttsFragment;
 
+    private long articleID;
     private Article mArticle;
     private ArticleDao mArticleDao;
 
@@ -111,8 +112,8 @@ public class ReadArticleActivity extends BaseActionBarActivity {
         setContentView(R.layout.article);
 
         Intent intent = getIntent();
-        long articleId = intent.getLongExtra(EXTRA_ID, -1);
-        Log.d(TAG, "onCreate() articleId=" + articleId);
+        articleID = intent.getLongExtra(EXTRA_ID, -1);
+        Log.d(TAG, "onCreate() articleId=" + articleID);
         if(intent.hasExtra(EXTRA_LIST_FAVORITES)) {
             contextFavorites = intent.getBooleanExtra(EXTRA_LIST_FAVORITES, false);
         }
@@ -122,14 +123,13 @@ public class ReadArticleActivity extends BaseActionBarActivity {
 
         DaoSession session = DbConnection.getSession();
         mArticleDao = session.getArticleDao();
-        mArticle = mArticleDao.queryBuilder()
-                .where(ArticleDao.Properties.Id.eq(articleId)).build().unique();
+        mArticle = getArticle(articleID);
 
         // article is loaded - update menu
         invalidateOptionsMenu();
 
         if(mArticle == null) {
-            Log.e(TAG, "onCreate() Did not find article with articleId=" + articleId + ". Thus we" +
+            Log.e(TAG, "onCreate() Did not find article with articleId=" + articleID + ". Thus we" +
                     " are not able to create this activity. Finish.");
             finish();
             return;
@@ -638,8 +638,11 @@ public class ReadArticleActivity extends BaseActionBarActivity {
         if(loadingFinished && mArticle != null) {
             cancelPositionRestoration();
 
-            mArticle.setArticleProgress(getReadingPosition());
-            mArticleDao.update(mArticle);
+            Article article = getArticle(articleID);
+            if(article != null) {
+                article.setArticleProgress(getReadingPosition());
+                mArticleDao.update(article);
+            }
         }
         super.onStop();
     }
@@ -788,6 +791,10 @@ public class ReadArticleActivity extends BaseActionBarActivity {
         return result;
     }
 
+
+    private Article getArticle(long articleID) {
+        return mArticleDao.queryBuilder().where(ArticleDao.Properties.Id.eq(articleID)).unique();
+    }
 
     private Long getAdjacentArticle(boolean previous) {
         QueryBuilder<Article> qb = mArticleDao.queryBuilder();

--- a/app/src/main/java/fr/gaulupeau/apps/Poche/ui/ReadArticleActivity.java
+++ b/app/src/main/java/fr/gaulupeau/apps/Poche/ui/ReadArticleActivity.java
@@ -73,7 +73,6 @@ public class ReadArticleActivity extends BaseActionBarActivity {
     private LinearLayout bottomTools;
     private View hrBar;
     private TtsFragment ttsFragment;
-    private MenuItem menuTTS;
 
     private Article mArticle;
     private ArticleDao mArticleDao;
@@ -125,6 +124,9 @@ public class ReadArticleActivity extends BaseActionBarActivity {
         mArticleDao = session.getArticleDao();
         mArticle = mArticleDao.queryBuilder()
                 .where(ArticleDao.Properties.Id.eq(articleId)).build().unique();
+
+        // article is loaded - update menu
+        invalidateOptionsMenu();
 
         if(mArticle == null) {
             Log.e(TAG, "onCreate() Did not find article with articleId=" + articleId + ". Thus we" +
@@ -569,22 +571,24 @@ public class ReadArticleActivity extends BaseActionBarActivity {
         MenuInflater inflater = getMenuInflater();
         inflater.inflate(R.menu.option_article, menu);
 
-        boolean unread = mArticle.getArchive() != null && !mArticle.getArchive();
+        if(mArticle != null) {
+            boolean unread = mArticle.getArchive() != null && !mArticle.getArchive();
 
-        MenuItem markReadItem = menu.findItem(R.id.menuArticleMarkAsRead);
-        markReadItem.setTitle(unread ? R.string.btnMarkRead : R.string.btnMarkUnread);
+            MenuItem markReadItem = menu.findItem(R.id.menuArticleMarkAsRead);
+            markReadItem.setTitle(unread ? R.string.btnMarkRead : R.string.btnMarkUnread);
 
-        boolean favorite = mArticle.getFavorite() != null && mArticle.getFavorite();
+            boolean favorite = mArticle.getFavorite() != null && mArticle.getFavorite();
 
-        MenuItem toggleFavoriteItem = menu.findItem(R.id.menuArticleToggleFavorite);
-        toggleFavoriteItem.setTitle(
-                favorite ? R.string.remove_from_favorites : R.string.add_to_favorites);
-        toggleFavoriteItem.setIcon(getIcon(favorite
-                        ? R.drawable.ic_star_white_24dp
-                        : R.drawable.ic_star_border_white_24dp, null)
-        );
+            MenuItem toggleFavoriteItem = menu.findItem(R.id.menuArticleToggleFavorite);
+            toggleFavoriteItem.setTitle(
+                    favorite ? R.string.remove_from_favorites : R.string.add_to_favorites);
+            toggleFavoriteItem.setIcon(getIcon(favorite
+                    ? R.drawable.ic_star_white_24dp
+                    : R.drawable.ic_star_border_white_24dp, null)
+            );
+        }
 
-        menuTTS = menu.findItem(R.id.menuTTS);
+        MenuItem menuTTS = menu.findItem(R.id.menuTTS);
         menuTTS.setChecked(ttsFragment != null);
 
         return true;
@@ -778,9 +782,9 @@ public class ReadArticleActivity extends BaseActionBarActivity {
             settings.setTtsVisible(false);
             result = false;
         }
-        if (menuTTS != null) {
-            menuTTS.setChecked(ttsFragment != null);
-        }
+
+        invalidateOptionsMenu();
+
         return result;
     }
 

